src/js/protocols/stm32.js

```diff
@@ -85,49 +88,120 @@ STM32_protocol.prototype.connect = function (port, baud, hex, options, callback)
-                            setTimeout(function() {
-                                // refresh device list
-                                PortHandler.check_usb_devices(function(dfu_available) {
-                                    if(dfu_available) {
-                                        STM32DFU.connect(usbDevices, hex, options);
-                                    } else {
-                                        serial.connect(port, {bitrate: self.baud, parityBit: 'even', stopBits: 'one'}, function (openInfo) {
-                                            if (openInfo) {
-                                                self.initialize();
-                                            } else {
-                                                GUI.connect_lock = false;
-                                                GUI.log(i18n.getMessage('serialPortOpenFail'));
-                                            }
-                                        });
-                                    }
-                                });
-                            }, 1000);
+                            setTimeout(startFlashing, 1000);
                         } else {
                             GUI.connect_lock = false;
                         }
                     });
                 });
+            });
+        };
+
+        var onConnectHandler = function () {
+
+            GUI.log(i18n.getMessage('apiVersionReceived', [CONFIG.apiVersion]));
+
+            if (semver.lt(CONFIG.apiVersion, "1.42.0")) {
+
+                self.msp_connector.disconnect(function (disconnectionResult) {
+
+                    // need some time for the port to be closed, serial port does not open if tried immediately
+                    setTimeout(legacyRebootAndFlash, 500);
+                });
             } else {
-                GUI.log(i18n.getMessage('serialPortOpenFail'));
+                console.log('Looking for capabilities via MSP');
+
+                MSP.send_message(MSPCodes.MSP_BOARD_INFO, false, false, function () {
+                    var rebootMode = 0; // FIRMWARE
+                    if (bit_check(CONFIG.targetCapabilities, FC.TARGET_CAPABILITIES_FLAGS.HAS_FLASH_BOOTLOADER)) {
+                        // Board has flash bootloader
+                        GUI.log(i18n.getMessage('deviceRebooting_flashBootloader'));
+                        console.log('flash bootloader detected');
+                        rebootMode = 4; // MSP_REBOOT_BOOTLOADER_FLASH
+                    } else {
+                        GUI.log(i18n.getMessage('deviceRebooting_romBootloader'));
+                        console.log('no flash bootloader detected');
+                        rebootMode = 1; // MSP_REBOOT_BOOTLOADER_ROM;
+                    }
+
+                    var buffer = [];
+                    buffer.push8(rebootMode);
+                    MSP.send_message(MSPCodes.MSP_SET_REBOOT, buffer, function() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['MSP' is not defined.](https://app.codacy.com/gh/emuflight/EmuConfigurator/pullRequest?prid=5626707)